test/Microsoft.AspNetCore.Razor.Runtime.Test/Runtime/TagHelpers/TagHelperDescriptorFactoryTest.cs

```diff
@@ -19,6 +19,152 @@ public class TagHelperDescriptorFactoryTest
+                    { "[name='value'", error(Resources.FormatTagHelperDescriptorFactory_CouldNotFindMatchingEndBrace("[name='value'")) },
+                    { "[name ", error(Resources.FormatTagHelperDescriptorFactory_CouldNotFindMatchingEndBrace("[name ")) },
+                    { "[name=value ", error(Resources.FormatTagHelperDescriptorFactory_CouldNotFindMatchingEndBrace("[name=value ")) },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(RequiredAttributeParserErrorData))]
+        public void RequiredAttributeParser_ParsesRequiredAttributesAndLogsErrorCorrectly(
+            string requiredAttributes,
+            RazorError expectedError)
+        {
+            // Arrange
+            var parser = new TagHelperDescriptorFactory.RequiredAttributeParser(requiredAttributes);
+            var errorSink = new ErrorSink();
+            IEnumerable<TagHelperRequiredAttributeDescriptor> descriptors;
+
+            // Act
+            var parsedCorrectly = parser.TryParse(errorSink, out descriptors);
+
+            // Assert
+            Assert.False(parsedCorrectly);
+            Assert.Null(descriptors);
+            var error = Assert.Single(errorSink.Errors);
+            Assert.Equal(expectedError, error);
+        }
+
+        public static TheoryData RequiredAttributeParserData
+        {
+            get
+            {
+                Func<string, TagHelperRequiredAttributeNameComparison, TagHelperRequiredAttributeDescriptor> plain =
+                    (name, nameComparison) => new TagHelperRequiredAttributeDescriptor
+                    {
+                        Name = name,
+                        NameComparison = nameComparison
+                    };
+                Func<string, string, TagHelperRequiredAttributeValueComparison, TagHelperRequiredAttributeDescriptor> css =
+                    (name, value, valueComparison) => new TagHelperRequiredAttributeDescriptor
+                    {
+                        Name = name,
+                        NameComparison = TagHelperRequiredAttributeNameComparison.FullMatch,
+                        Value = value,
+                        ValueComparison = valueComparison,
+                    };
+
+                return new TheoryData<string, IEnumerable<TagHelperRequiredAttributeDescriptor>>
+                {
+                    { null, Enumerable.Empty<TagHelperRequiredAttributeDescriptor>() },
+                    { string.Empty, Enumerable.Empty<TagHelperRequiredAttributeDescriptor>() },
+                    { "name", new[] { plain("name", TagHelperRequiredAttributeNameComparison.FullMatch) } },
+                    { "name-*", new[] { plain("name-", TagHelperRequiredAttributeNameComparison.PrefixMatch) } },
+                    { "  name-*   ", new[] { plain("name-", TagHelperRequiredAttributeNameComparison.PrefixMatch) } },
```

Comment:
Errors about whitespace after `']'` look a bit more bizarre given this is successful.